File: app/src/main/java/com/liaudev/dicodingbpaa/ui/splash/SplashActivity.java
package com.liaudev.dicodingbpaa.ui.splash;


import android.animation.Animator;
import android.animation.AnimatorSet;
import android.animation.ObjectAnimator;
import android.app.ActivityOptions;
import android.content.Intent;
import android.os.Build;
import android.os.Bundle;
import android.util.Pair;
import android.view.View;
import android.view.animation.AccelerateInterpolator;
import android.view.animation.BounceInterpolator;

import com.liaudev.dicodingbpaa.App;
import com.liaudev.dicodingbpaa.R;
import com.liaudev.dicodingbpaa.databinding.ActivitySplashBinding;
import com.liaudev.dicodingbpaa.ui.BaseActivity;
import com.liaudev.dicodingbpaa.ui.login.LoginActivity;
import com.liaudev.dicodingbpaa.ui.main.MainActivity;


/**
 * Created by Budiliauw87 on 2022-05-14.
 * budiliauw87.github.io
 * Budiliauw87@gmail.com
 */

public class SplashActivity extends BaseActivity {
    private ActivitySplashBinding binding;
    private boolean mShouldFinish = false;
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        binding = ActivitySplashBinding.inflate(getLayoutInflater());
        setContentView(binding.getRoot());
        //change statusbar color
        updateStatusBarColor(getColor(R.color.transparent));
        //animate
        ObjectAnimator logo = ObjectAnimator.ofFloat(binding.imgLogo,View.TRANSLATION_Y,-150f,0).setDuration(2000);
        ObjectAnimator objTitle = ObjectAnimator.ofFloat(binding.tvTitle,View.ALPHA,1f).setDuration(500);
        ObjectAnimator objSubTitle = ObjectAnimator.ofFloat(binding.tvSubTitle,View.ALPHA,1f).setDuration(500);
        ObjectAnimator objVersion = ObjectAnimator.ofFloat(binding.tvVersion,View.TRANSLATION_Y,150f,0).setDuration(500);
        logo.setInterpolator(new BounceInterpolator());
        logo.addListener(new Animator.AnimatorListener() {
            @Override
            public void onAnimationStart(Animator animator) {

            }

            @Override
            public void onAnimationEnd(Animator animator) {
                //to activity
                Intent intent;
                if(App.getInstance().isLogin()){
                    intent = new Intent(SplashActivity.this, MainActivity.class);
                    startActivity(intent);
                }else{
                    intent = new Intent(SplashActivity.this, LoginActivity.class);
                    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
                        Pair[] pairs = new Pair[3];
                        pairs[0] = new Pair<View, String>(binding.imgLogo, "logo_image");
                        pairs[1] = new Pair<View, String>(binding.tvTitle, "logo_title");
                        pairs[2] = new Pair<View, String>(binding.tvSubTitle, "logo_subtitle");

                        ActivityOptions activityOptions = ActivityOptions.makeSceneTransitionAnimation(SplashActivity.this, pairs);
                        startActivity(intent, activityOptions.toBundle());
                    }else{
                        startActivity(intent);
                    }
                }

                mShouldFinish = true;

            }

            @Override
            public void onAnimationCancel(Animator animator) {

            }

            @Override
            public void onAnimationRepeat(Animator animator) {

            }
        });
        logo.start();
        AnimatorSet animatorSet = new AnimatorSet();
        animatorSet.setDuration(1000);
        animatorSet.playTogether(objTitle,objSubTitle,objVersion);
        animatorSet.setInterpolator(new AccelerateInterpolator());
        animatorSet.start();
    }

    @Override
    public void onStop() {
        super.onStop();
        if(mShouldFinish)
            finish();
    }
}

File: app/src/main/java/com/liaudev/storydicoding/ui/splash/SplashActivity.kt
package com.liaudev.storydicoding.ui.splash

import android.animation.AnimatorSet
import android.animation.ObjectAnimator
import android.content.Intent
import android.os.Build
import android.os.Bundle
import android.util.Log
import android.view.View
import android.view.WindowInsets
import android.view.WindowManager
import android.view.animation.BounceInterpolator
import androidx.activity.viewModels
import androidx.appcompat.app.AppCompatActivity
import androidx.core.animation.addListener
import com.liaudev.storydicoding.databinding.ActivitySplashBinding
import com.liaudev.storydicoding.di.ViewModelFactory
import com.liaudev.storydicoding.ui.login.LoginActivity
import com.liaudev.storydicoding.ui.main.MainActivity

/**
 * Created by Budiliauw87 on 2022-10-19.
 * budiliauw87.github.io
 * Budiliauw87@gmail.com
 */
class SplashActivity : AppCompatActivity() {
    private lateinit var binding: ActivitySplashBinding
    private lateinit var factory: ViewModelFactory
    private val viewModel: SplashViewModel by viewModels { factory }
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivitySplashBinding.inflate(layoutInflater)
        setContentView(binding.root)
        @Suppress("DEPRECATION")
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
            window.insetsController?.hide(WindowInsets.Type.statusBars())
        } else {
            window.setFlags(
                WindowManager.LayoutParams.FLAG_FULLSCREEN,
                WindowManager.LayoutParams.FLAG_FULLSCREEN
            )
        }
        supportActionBar?.hide()
        factory = ViewModelFactory.getInstance(this)
        startAnimation()
    }

    private fun startAnimation() {

        val objTitle = ObjectAnimator.ofFloat(binding.tvTitle, View.ALPHA, 1f).setDuration(500)
        val objSubTitle =
            ObjectAnimator.ofFloat(binding.tvSubTitle, View.ALPHA, 1f).setDuration(500)
        val objVersion =
            ObjectAnimator.ofFloat(binding.tvVersion, View.TRANSLATION_Y, 150f, 0f).setDuration(500)
        val logo =
            ObjectAnimator.ofFloat(binding.imgLogo, View.TRANSLATION_Y, -150f, 0f).setDuration(2000)
        logo.interpolator = BounceInterpolator()
        AnimatorSet().apply {
            playSequentially(logo, objTitle, objSubTitle, objVersion)
            addListener(onEnd = {
                viewModel.getUser().observe(this@SplashActivity) { user ->
                    Log.e("SplashActivity", user.toString())
                    val intent = if (user.isLogin && user.token != "") {
                        Intent(this@SplashActivity, MainActivity::class.java)
                    } else {
                        Intent(this@SplashActivity, LoginActivity::class.java)
                    }
                    startActivity(intent)
                    finish()
                }

            })
        }.start()

    }
}